File: agent/nodes/planAgent.js
const { ChatOpenAI } = require('@langchain/openai');

const llm = new ChatOpenAI({ model: 'gpt-4o', temperature: 0.2 });

/**
 * PLAN AGENT node — read the Jira ticket and decide what files to build.
 * @param {import('../state').AgentStateAnnotation} state
 */
async function planAgentNode(state) {
  console.log('\n🧠 NODE: planAgent — planning implementation...');

  const { currentTicket } = state;
  if (!currentTicket) {
    console.log('  ⚠️  No current ticket — skipping.');
    return { ...state, plan: null };
  }

  const summary = currentTicket.fields?.summary ?? '';
  const description =
    typeof currentTicket.fields?.description === 'object'
      ? JSON.stringify(currentTicket.fields.description)
      : (currentTicket.fields?.description ?? 'No description');
  const labels = (currentTicket.fields?.labels ?? []).join(', ');

  const prompt = `You are a senior software architect planning a feature for a Student Management System built with Next.js 14 (App Router), TypeScript, and Tailwind CSS.

Jira Ticket:
- Key: ${currentTicket.key}
- Summary: ${summary}
- Labels: ${labels || 'none'}
- Description: ${description}

Based on this ticket, produce a JSON implementation plan with:
1. workType: "frontend" | "backend" | "fullstack"
2. files: array of { path: string, description: string } — every file that needs to be created or modified

Rules:
- Frontend files: app/[page]/page.tsx, components/[Name].tsx
- Backend files: app/api/[resource]/route.ts, lib/[name].ts
- Always follow App Router conventions (app/ directory)
- Be specific about file paths

Respond ONLY with valid JSON, no markdown, no explanation.

Example:
{
  "workType": "fullstack",
  "files": [
    { "path": "app/students/[id]/page.tsx", "description": "Student detail page showing all info and grades" },
    { "path": "app/api/students/[id]/grades/route.ts", "description": "GET endpoint to fetch grades for a student" }
  ]
}`;

  try {
    const response = await llm.invoke(prompt);
    const raw = response.content.toString().trim();
    const plan = JSON.parse(raw.replace(/```json|```/g, '').trim());

    console.log(`  ✅ Plan: ${plan.workType} — ${plan.files.length} file(s) to build`);
    plan.files.forEach((f) => console.log(`    - ${f.path}`));

    return { ...state, plan };
  } catch (err) {
    console.error('  ❌ planAgent failed:', err.message);
    const fallbackPlan = {
      workType: labels.includes('backend') ? 'backend' : labels.includes('frontend') ? 'frontend' : 'fullstack',
      files: [],
      error: err.message,
    };
    return { ...state, plan: fallbackPlan };
  }
}

module.exports = { planAgentNode };

File: python-backend/nodes/plan_agent.py
import json
from langchain_openai import ChatOpenAI
from state import AgentState
from tools.logger import jira_log

_llm = None


def _get_llm() -> ChatOpenAI:
    global _llm
    if _llm is None:
        _llm = ChatOpenAI(model="gpt-4o", temperature=0.2)
    return _llm


async def plan_agent_node(state: AgentState) -> AgentState:
    print("\n🧠 NODE: planAgent — planning implementation...")
    ticket = state.get("current_ticket")

    if not ticket:
        print("  ⚠️  No current ticket — skipping.")
        return state

    key = ticket["key"]
    summary = ticket.get("fields", {}).get("summary", "")
    description = ticket.get("fields", {}).get("description", "") or ""

    await jira_log(key, "🧠 PlanAgent Started")

    prompt = f"""You are a senior software architect planning features for a Next.js 14 student management system.

Ticket: {ticket['key']} — {summary}
Description: {description}

The project has these existing routes and components:
- app/page.tsx (dashboard)
- app/students/page.tsx, app/students/[id]/page.tsx
- app/courses/page.tsx
- app/grades/page.tsx
- app/attendance/page.tsx
- components/DataTable.tsx, components/Navbar.tsx, components/Footer.tsx
- app/api/students/, app/api/courses/, app/api/grades/, app/api/attendance/
- lib/db.ts (getDb returns {{students, courses, grades}})
- lib/attendance.ts (attendanceStore)

Respond with ONLY valid JSON (no markdown):
{{
  "workType": "frontend" | "backend" | "fullstack",
  "files": ["path/to/file1.tsx", "path/to/file2.ts"],
  "description": "brief plan"
}}

Rules:
- "frontend" for UI-only changes (pages, components)
- "backend" for API routes only
- "fullstack" for both
- List only files that need to be CREATED or MODIFIED
- Keep scope minimal — only what the ticket requires"""

    try:
        response = await _get_llm().ainvoke(prompt)
        content = response.content.strip()
        for prefix in ["```json", "```"]:
            content = content.removeprefix(prefix)
        content = content.removesuffix("```").strip()
        plan = json.loads(content)
        work_type = plan.get("workType", "frontend")
        files = plan.get("files", [])
        print(f"  ✅ Plan: {work_type} — {len(files)} file(s) to build")
        for f in files:
            print(f"    - {f}")
        await jira_log(key, f"🧠 PlanAgent Completed\nType: {work_type}\nFiles: {', '.join(files)}")
        return {**state, "plan": plan}
    except Exception as e:
        print(f"  ❌ Plan failed: {e}")
        await jira_log(key, f"❌ PlanAgent Failed\nError: {e}")
        fallback = {"workType": "frontend", "files": [], "description": str(e)}
        return {**state, "plan": fallback, "error": str(e)}
